### PR TITLE
Workflow Reset XDC implementation

### PR DESCRIPTION
**Pull request by longquanzheng**
Finish all unit test.
I am implementing integration test

### CODE REVIEW

File: common/persistence/dataInterfaces.go [hunk 1]
```diff
@@ -768,6 +768,10 @@ type (
 		Condition int64
 		RangeID   int64
 
+		// for fork run (we need to make sure the forkRun hasn't been deleted)
+		ForkRunID          string
+		ForkRunNextEventID int64
```
Comment (wxing1292) [inline common/persistence/dataInterfaces.go]: probably should also include ForkRunLastWriteVersion

File: service/history/historyReplicator.go [hunk 2]
```diff
@@ -60,6 +60,7 @@ type (
 		clusterMetadata   cluster.Metadata
 		metricsClient     metrics.Client
 		logger            bark.Logger
+		resetor           *workflowResetor
```
Comment (wxing1292) [inline service/history/historyReplicator.go]: this should be a interface

File: service/history/historyReplicator.go [hunk 3]
```diff
@@ -678,6 +690,177 @@ func (r *historyReplicator) ApplyReplicationTask(ctx context.Context, context wo
...
+		return nil, errUnknownReplicationTask
+	}
+	firstEvent := newRunHistory[0]
+	if firstEvent.DecisionTaskFailedEventAttributes.GetCause() != workflow.DecisionTaskFailedCauseResetWorkflow {
+		return nil, errUnknownReplicationTask
+	}
+	return firstEvent.DecisionTaskFailedEventAttributes, nil
+}
+
+func (r *historyReplicator) replicateResetEvent(forkMutableState mutableState, forkExecution *workflow.WorkflowExecution, newRunHistory []*workflow.HistoryEvent) (newMsBuilder mutableState, transferTasks, timerTasks []persistence.Task, retError error) {
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply service/history/historyReplicator.go]: OK, will move to workflowResetor as we discussed offline

File: service/history/historyReplicator.go [hunk 4]
```diff
@@ -678,6 +690,177 @@ func (r *historyReplicator) ApplyReplicationTask(ctx context.Context, context wo
 	return err
 }
 
+func (r *historyReplicator) ApplyResetEvent(ctx context.Context, request *h.ReplicateEventsRequest, domainID, requestID string, currExecution workflow.WorkflowExecution, currContext workflowExecutionContext, sBuilder stateBuilder, currMutableState mutableState) (retError error) {
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply service/history/historyReplicator.go]: OK, will move to workflowResetor as we discussed offline

File: service/history/historyReplicator.go (hunk 2 above)
Comment (longquanzheng) [reply service/history/historyReplicator.go]: OK

File: common/persistence/dataInterfaces.go (hunk 1 above)
Comment (longquanzheng) [reply common/persistence/dataInterfaces.go]: OK

Comment (longquanzheng) [reply common/persistence/dataInterfaces.go]: Confirmed offline that what we need is PrevRunVersion/PrevRunState for the current run to assert on current workflow record

**Comment by longquanzheng** [comment]
@wxing1292 @samarabbas  Do you have time to take a look?

File: common/persistence/cassandra/cassandraPersistence.go [hunk 5]
```diff
@@ -1872,25 +1870,72 @@ func (d *cassandraPersistence) ResetWorkflowExecution(request *p.InternalResetWo
...
+			insertExecutionInfo.DomainID,
+			insertExecutionInfo.WorkflowID,
+			permanentRunID,
+			defaultVisibilityTimestamp,
+			rowTypeExecutionTaskID,
+			currExecutionInfo.RunID,
+		)
+	} else {
+		// NOTE: workflow_state is either 1 or 2, there is no 0(created)
+		prevRunState := 2
```
Comment (wxing1292) [inline common/persistence/cassandra/cassandraPersistence.go]: i think this value should be passed in, btw, plz use WorkflowStateRunning or WorkflowStateCompleted

File: service/history/workflowResetor.go [hunk 6]
```diff
@@ -223,24 +255,33 @@ func validateResetWorkflowAfterReplay(newMutableState mutableState) (retError er
...
 	for _, ai := range msBuilder.GetPendingActivityInfos() {
 		if ai.StartedID != common.EmptyEventID {
 			// this means the activity has started but not completed, we need to fail the activity
 			request := getRespondActivityTaskFailedRequestFromActivity(ai, "workflowReset")
 			if msBuilder.AddActivityTaskFailedEvent(ai.ScheduleID, ai.StartedID, request) == nil {
 				// Unable to add ActivityTaskFailed event to history
-				return nil, &workflow.InternalServiceError{Message: "Unable to add ActivityTaskFailed event to mutableState."}
+				return &workflow.InternalServiceError{Message: "Unable to add ActivityTaskFailed event to mutableState."}
 			}
 			continue
```
Comment (wxing1292) [inline service/history/workflowResetor.go]: this continue does nothing?

File: service/history/workflowResetor.go [hunk 7]
```diff
@@ -353,51 +397,42 @@ func (w *workflowResetor) setEventIDsWithHistory(msBuilder mutableState) int64 {
...
 }
 
-func (w *workflowResetor) generateReplicationTasksForReset(terminateCurr bool, currMutableState, newMutableState mutableState, domainEntry *cache.DomainCacheEntry) []persistence.Task {
+func (w *workflowResetorImpl) generateReplicationTasksForReset(terminateCurr bool, currMutableState, newMutableState mutableState, domainEntry *cache.DomainCacheEntry) []persistence.Task {
 	var repTasks []persistence.Task
 	if newMutableState.GetReplicationState() != nil {
 		if terminateCurr {
-			firstEventIDForNew := w.setEventIDsWithHistory(newMutableState)
+			// we will 2 replication tasks for this case
 			firstEventIDForCurr := w.setEventIDsWithHistory(currMutableState)
```
Comment (wxing1292) [inline service/history/workflowResetor.go]: we generate will 2 replication tasks for this case

File: service/history/workflowResetor.go [hunk 8]
```diff
@@ -558,3 +594,212 @@ func validateLastBatchOfReset(lastBatch []*workflow.HistoryEvent) error {
...
+	if currentRunID == forkAttr.GetForkRunId() {
+		currMutableState = forkMutableState
+		currContext = forkContext
+	} else {
+		var currRelease releaseWorkflowExecutionFunc
+		currExecution := workflow.WorkflowExecution{
+			WorkflowId: forkExecution.WorkflowId,
+			RunId:      common.StringPtr(currentRunID),
+		}
+		currContext, currRelease, retError = w.eng.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, currExecution)
```
Comment (wxing1292) [inline service/history/workflowResetor.go]: maybe this should be currErr?

line 631 should probably be forkErr?

File: common/persistence/cassandra/cassandraPersistence.go (hunk 5 above)
Comment (longquanzheng) [reply common/persistence/cassandra/cassandraPersistence.go]: OK.

File: service/history/workflowResetor.go (hunk 6 above)
Comment (longquanzheng) [reply service/history/workflowResetor.go]: It is useless, I will remove it.

File: service/history/workflowResetor.go (hunk 7 above)
Comment (longquanzheng) [reply service/history/workflowResetor.go]: fixed

File: service/history/workflowResetor.go (hunk 8 above)
Comment (longquanzheng) [reply service/history/workflowResetor.go]: I am OK to do that. I guess the benefit of doing that is with currErr, we can avoid releasing forkMutableState without clear the cache. I don't quite know what do you want to achieve to use forkErr. But that's also fine to be consistent.

**Pull request merged**